constraint.go

```diff
@@ -0,0 +1,358 @@
+package semver
+
+import (
+	"errors"
+	"fmt"
+)
+
+var noneErr = errors.New("The 'None' constraint admits no versions.")
+
+type Constraint interface {
+	// Admits checks that a version satisfies the constraint. If it does not,
+	// an error is returned indcating the problem; if it does, the error is nil.
+	Admits(v *Version) error
+
+	// Intersect computes the intersection between the receiving Constraint and
+	// passed Constraint, and returns a new Constraint representing the result.
+	Intersect(Constraint) Constraint
+
+	// AdmitsAny returns a bool indicating whether there exists any version that
+	// can satisfy the Constraint.
+	AdmitsAny() bool
```

Comment:
It reports whether it's possible that *any* version could satisfy the constraint. Basically, it's asking of the constraint, "are you the empty set?". This is important for a solver - in a situation where project `A` has a dependency on `C` with constraint `x`, and then project `B` comes along with a dependency on `B` using constraint `z`, the way we determine if it's acceptable to pull in `B` is by computing the intersection of the constraint (`x ∩ z`), and seeing if the result of that `AdmitsAny()`. Even if it does, of course, that doesn't guarantee we'll be able to find version that would satisfy the constraint - but we figure that out later. What matters, [at that moment in the algorithm](https://github.com/sdboyer/vsolver/blob/master/solver.go#L223) (note that I need to rename that method - it's intersection, not union) is that if the intersection DOESN'T allow anything, then we know we have to back off and try something else.

I'm not super-happy with the implementation right now - the way dart does it, the method takes a constraint argument, performs an intersection and then reports if the result of the intersection admits anything. Once I finish getting the basic impl of `unionConstraint` and `rangeConstraint` done, I may revisit this.